bndtools.m2e/src/bndtools/m2e/MavenWorkspaceRepository.java

```diff
@@ -0,0 +1,336 @@
+    private boolean inited = false;
+
+    private final Map<String,Entry<IMavenProjectFacade,MavenProject>> bsnMap = new HashMap<>();
+
+    @Override
+    public Map<Requirement,Collection<Capability>> findProviders(Collection< ? extends Requirement> requirements) {
+        return null;
+    }
+
+    @Override
+    public boolean refresh() throws Exception {
+        return false;
+    }
+
+    @Override
+    public File getRoot() throws Exception {
+        return null;
+    }
+
+    @Override
+    public PutResult put(InputStream stream, PutOptions options) throws Exception {
+        return null;
+    }
+
+    @Override
+    public File get(final String bsn, final Version version, Map<String,String> properties, final DownloadListener... listeners) throws Exception {
+        if (!inited) {
+            init();
+        }
+
+        final Entry<IMavenProjectFacade,MavenProject> entry = bsnMap.get(bsn);
+
+        if (entry == null)
+            return null;
+
+        final MavenProject mavenProject = entry.getValue();
+
+        // add the eclipse project that this comes from so we can look it up in the launch
+        // see bndtools.launch.BndDependencySourceContainer.createSourceContainers()
+        final String projectName = entry.getKey().getProject().getName();
+        properties.put("sourceProjectName", projectName);
+
+        final File bundleFile = new File(mavenProject.getBuild().getDirectory(), mavenProject.getBuild().getFinalName() + ".jar");
+
+        if (listeners == null || listeners.length == 0) {
+            String errorIfAny = createBundleFileIfNeeded(bundleFile, bsn, version, mavenProject, new NullProgressMonitor());
+
+            if (errorIfAny != null) {
+                return null;
+            }
+
+            return bundleFile;
+        }
+
+        new Job(getName() + " getting " + bsn) {
+            @Override
+            public IStatus run(IProgressMonitor monitor) {
+                String errorIfAny = createBundleFileIfNeeded(bundleFile, bsn, version, mavenProject, monitor);
+
+                for (DownloadListener listener : listeners) {
+                    try {
+                        if (errorIfAny == null) {
+                            listener.success(bundleFile);
+                        } else {
+                            listener.failure(bundleFile, errorIfAny);
+                        }
+                    } catch (Throwable t) {
+                        logger.logError("Download listener error", t);
+                    }
+                }
+
+                return Status.OK_STATUS;
+            }
+        }.schedule();
+
+        return bundleFile;
+    }
+
+    private String createBundleFileIfNeeded(File bundleFile, String bsn, Version version, MavenProject mavenProject, IProgressMonitor monitor) {
+        String failureReason = null;
+
+        if (isBundleFileOutOfDate(bundleFile, mavenProject)) {
+            try {
+                MavenUtil.execJarMojo(mavenProject, monitor);
```

Comment:
To be honest I feel that the jar will never be out of date as far as a normal user workflow in eclipse (with eclipse "Build Automatically" turned on).  The reason I had this in here is for the "abnormal case" where for some reason the bnd-process mojo ran but the BndConfigurator's build participant didn't execute (perhaps user has build-automatically turn off?) and we need to create the jar 'just-in-time'.  

As for triggering the normal eclipse build inside the MavenWorkspaceRepository, the problem there is that the "jar:jar" mojo execution happens on a async WorkspaceJob, so its hard to tell when that has finished.  

If we want to simplify this class for now, I could remove the building of the jar from the Repo itself and just make it simply grab the jar that is there always and assume that the Eclipse builder (and maven m2e participants) have done their job properly.